### PR TITLE
Add a lint for `slice.iter().cloned().collect()`

### LINKED ISSUE #1292: lint unnecessary iteration
rustc code contained `slice.iter().cloned().collect()`, which should be `slice.to_owned()`.

is probably just another extension to https://github.com/Manishearth/rust-clippy/blob/master/clippy_lints/src/methods.rs#L502

### PR DESCRIPTION
**Pull request by F001**
If one uses `slice.iter().cloned().collect()` to create a new `Vec`,
it should be `slice.to_owned()`.

Fix #1292

### CODE REVIEW

File: clippy_lints/src/methods.rs [hunk 1]
```diff
@@ -494,13 +494,33 @@ declare_lint! {
...
 /// s.push_str(&def));
 /// ```
-
 declare_lint! {
     pub STRING_EXTEND_CHARS,
     Warn,
     "using `x.extend(s.chars())` where s is a `&str` or `String`"
 }
 
+/// **What it does:** Checks for the use of `.cloned().collect()` on slice to create a Vec.
```
Comment (mcarton) [inline clippy_lints/src/methods.rs]: Either “a vector” or “a `Vec`”.

File: clippy_lints/src/methods.rs [hunk 2]
```diff
@@ -494,13 +494,33 @@ declare_lint! {
...
+/// **Why is this bad?** `.to_owned()` is clearer
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// let s = [1,2,3,4,5];
+/// let s2 : Vec<isize> = s.iter().cloned().collect();
+/// ```
+/// The correct use would be:
```
Comment (mcarton) [inline clippy_lints/src/methods.rs]: s/The correct/A better

“The correct” implies it's wrong, but it's only bad :smile:

File: clippy_lints/src/methods.rs [hunk 3]
```diff
@@ -879,6 +902,17 @@ fn lint_cstring_as_ptr(cx: &LateContext, expr: &hir::Expr, new: &hir::Expr, unwr
...
 }
 
+fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir::Expr]) {
+    if match_type(cx, cx.tables.expr_ty(expr), &paths::VEC) &&
+       derefs_to_slice(cx, &iter_args[0], cx.tables.expr_ty(&iter_args[0])).is_some() {
+        span_lint(cx,
+                  ITER_CLONED_COLLECT,
+                  expr.span,
+                  "called `cloned().collect()` on a slice to create a `Vec`. This is more succinctly expressed by \
+                  calling `to_owned(x)`");
```
Comment (mcarton) [inline clippy_lints/src/methods.rs]: As a bonus, do you mind trying to make a proper suggestion?

File: tests/ui/methods.rs [hunk 4]
```diff
@@ -689,3 +689,8 @@ fn temporary_cstring() {
 
 
 }
+
+fn iter_clone_collect() {
+    let v = [1,2,3,4,5];
+    let v2 : Vec<isize> = v.iter().cloned().collect();
+}
```
Comment (mcarton) [inline tests/ui/methods.rs]: Can you add a true-negative test with a conversion to a non-`Vec` as per https://github.com/Manishearth/rust-clippy/issues/1292#issuecomment-279205712?

**Comment by oli-obk** [comment]
thanks!

**Comment by malbarbo** [comment]
I think `to_vec` is more appropriated than `to_owned`.

**Pull request merged**